File: server/express/src/utils/errorHandler.ts
/**
 * Error Handling Utilities
 * 
 * This module provides centralized error handling for the Express application.
 * It includes middleware for catching and formatting errors in a consistent way.
 */

import { Request, Response, NextFunction } from 'express';
import { MongoError } from 'mongodb';

/**
 * Interface for standardized error responses
 */
interface ErrorResponse {
  success: false;
  error: {
    message: string;
    code?: string;
    details?: any;
  };
  timestamp: string;
}

/**
 * Global error handling middleware
 * 
 * This middleware catches all unhandled errors and returns a consistent
 * error response format. It should be the last middleware in the chain.
 * 
 * @param err - The error that was thrown
 * @param req - Express request object
 * @param res - Express response object
 * @param next - Express next function
 */
export function errorHandler(
  err: Error,
  req: Request,
  res: Response,
  next: NextFunction
): void {
  // Log the error for debugging purposes
  // In production, we recommend using a logging service
  console.error('Error occurred:', {
    message: err.message,
    stack: err.stack,
    url: req.url,
    method: req.method,
    timestamp: new Date().toISOString()
  });

  // Determine the appropriate HTTP status code and error message
  const errorResponse = createErrorResponse(err);
  
  // Send the error response
  res.status(errorResponse.statusCode).json({
    success: false,
    error: {
      message: errorResponse.message,
      code: errorResponse.code,
      details: errorResponse.details
    },
    timestamp: new Date().toISOString()
  } as ErrorResponse);
}

/**
 * Creates a standardized error response based on the error type
 * 
 * @param err - The error to process
 * @returns Object containing status code, message, and optional details
 */
function createErrorResponse(err: Error): {
  statusCode: number;
  message: string;
  code?: string;
  details?: any;
} {
  // Handle MongoDB-specific errors
  if (err instanceof MongoError) {
    return handleMongoError(err);
  }
  
  // Handle validation errors (you can extend this for custom validation)
  if (err.name === 'ValidationError') {
    return {
      statusCode: 400,
      message: 'Invalid input data',
      code: 'VALIDATION_ERROR',
      details: err.message
    };
  }
  
  // Handle JSON parsing errors
  if (err instanceof SyntaxError && 'body' in err) {
    return {
      statusCode: 400,
      message: 'Invalid JSON in request body',
      code: 'JSON_PARSE_ERROR'
    };
  }
  
  // Handle general application errors
  return {
    statusCode: 500,
    message: 'An unexpected error occurred',
    code: 'INTERNAL_SERVER_ERROR',
    details: process.env.NODE_ENV === 'development' ? err.message : undefined
  };
}

/**
 * Handles MongoDB-specific errors and returns appropriate responses
 * 
 * @param err - MongoDB error instance
 * @returns Object with status code and message
 */
function handleMongoError(err: MongoError): {
  statusCode: number;
  message: string;
  code: string;
  details?: any;
} {
  switch (err.code) {
    case 11000:
      // Duplicate key error
      return {
        statusCode: 409,
        message: 'Duplicate entry found',
        code: 'DUPLICATE_KEY_ERROR',
        details: 'A document with this data already exists'
      };
      
    case 121:
      // Document validation failed
      return {
        statusCode: 400,
        message: 'Document validation failed',
        code: 'DOCUMENT_VALIDATION_ERROR',
        details: err.message
      };
      
    default:
      // Generic MongoDB error
      return {
        statusCode: 500,
        message: 'Database operation failed',
        code: 'MONGODB_ERROR',
        details: process.env.NODE_ENV === 'development' ? err.message : undefined
      };
  }
}

/**
 * Async wrapper function for route handlers
 * 
 * This function wraps async route handlers to automatically catch
 * and forward any errors to the error handling middleware.
 * 
 * Usage:
 * app.get('/route', asyncHandler(async (req, res) => {
 *   // Your async code here
 * }));
 * 
 * @param fn - Async route handler function
 * @returns Express middleware function
 */
export function asyncHandler(
  fn: (req: Request, res: Response, next: NextFunction) => Promise<any>
) {
  return (req: Request, res: Response, next: NextFunction) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}

/**
 * Creates a standardized success response
 * 
 * @param data - The data to include in the response
 * @param message - Optional success message
 * @returns Standardized success response object
 */
export function createSuccessResponse(data: any, message?: string) {
  return {
    success: true,
    message: message || 'Operation completed successfully',
    data,
    timestamp: new Date().toISOString()
  };
}

/**
 * Validates that required fields are present in the request body
 * 
 * @param body - Request body object
 * @param requiredFields - Array of required field names
 * @throws Error if any required fields are missing
 */
export function validateRequiredFields(body: any, requiredFields: string[]): void {
  const missingFields = requiredFields.filter(field => 
    body[field] === undefined || body[field] === null || body[field] === ''
  );
  
  if (missingFields.length > 0) {
    throw new Error(`Missing required fields: ${missingFields.join(', ')}`);
  }
}

File: server/python/src/utils/errorHandler.py
from fastapi import Request
from fastapi.responses import JSONResponse
from pymongo.errors import PyMongoError, DuplicateKeyError, WriteError
from datetime import datetime, timezone
from typing import Any, Dict, Optional
from src.models.models import ErrorDetails, ErrorResponse, SuccessResponse, T


'''
Open to having a conversation about parity in the code. From my understanding exception handeling, validation errors and enforcement(Pydantic),
and error response formatting are all handled natively by FastAPI. So I don't believe I need to create the ValidationError
class, middleware, or exception handlers present in the TS code.

'''


'''
Creates a standardized success response.


Args:
    data (T): The data to include in the response.
    message (Optional[str]): An optional message to include.

Returns:
    SuccessResponse[T]: A standardized success response object.
    '''


# TODO: Verify the timestamp format is acceptable.
def create_success_response(data:T, message: Optional[str] = None) -> SuccessResponse[T]:
    return SuccessResponse(
        message=message or "Operation completed successfully.",
        data=data,
        timestamp=datetime.now(timezone.utc).isoformat() + "Z",
        
    )


'''
Creates a standardized error response.

Args:
    message (str): The error message.
    code (Optional[str]): An optional error code.
    details (Optional[Any]): Additional error details.

Returns:
    ErrorResponse: A standardized error response object.     

'''

# TODO: Verify the timestamp format is acceptable.
def create_error_response(message: str, code: Optional[str]=None, details: Optional[Any]=None) -> ErrorResponse:
    return ErrorResponse(
        message=message,
        error=ErrorDetails(
            message=message,
            code=code,
            details=details
        ),
        timestamp=datetime.now(timezone.utc).isoformat() + "Z",
    )



'''
This is interesting, I am not sure if this is worth explaining that compared to Node, you are 
not going to get exceptions thrown from MongoDB operations in the same way. You are not getting
error codes back from operations, you are getting exceptions. 
'''

def parse_mongo_exception(exc: Exception) -> dict:
    if isinstance(exc, DuplicateKeyError):
        return{
            "message": "Duplicate key error occurred.",
            "code": "DUPLICATE_KEY_ERROR",
            "details": "A document with the same key already exists.",
            "statusCode":409
        }
        
        # This is stating that the data that you are trying to implement is the wrong shape
        # for the schema implemented in MongoDB.
    elif isinstance(exc, WriteError):
        return{
            "message": "Document validation failed.",
            "code": "WRITE_ERROR",
            "details": str(exc),
            "statusCode":400
        }
        
    elif isinstance(exc, PyMongoError):
        return {
            "message" : "A database error occurred.",
            "code": "DATABASE_ERROR",
            "details": str(exc),
            "statusCode":500
        }
    return {
        "message": "An unknown error occurred.",
        "code": "UNKNOWN_ERROR",
        "details": str(exc),
        "statusCode": 500
    }

def register_error_handlers(app):

    @app.exception_handler(PyMongoError)
    async def mongo_exception_handler(request: Request, exc: PyMongoError):
        error_details = parse_mongo_exception(exc)
        return JSONResponse(
            status_code = error_details["statusCode"],
            content=create_error_response(
                message=error_details["message"],
                code=error_details["code"],
                details=error_details["details"]
            ).model_dump()
        )
    
    @app.exception_handler(Exception)
    async def generic_exception_handler(request: Request, exc: Exception):
        return JSONResponse(
            status_code=500,
            content=create_error_response(
                message=str(exc),
                code="INTERNAL_SERVER_ERROR",
                details=getattr(exc, 'detail', None) or getattr(exc, 'args', None)
            ).model_dump()
        )